File: jscomp/others/belt_MutableQueue.ml
(**************************************************************************)
(*                                                                        *)
(*                                 OCaml                                  *)
(*                                                                        *)
(*         Francois Pottier, projet Cristal, INRIA Rocquencourt           *)
(*                  Jeremie Dimino, Jane Street Europe                    *)
(*                                                                        *)
(*   Copyright 2002 Institut National de Recherche en Informatique et     *)
(*     en Automatique.                                                    *)
(*                                                                        *)
(*   All rights reserved.  This file is distributed under the terms of    *)
(*   the GNU Lesser General Public License version 2.1, with the          *)
(*   special exception on linking described in the file LICENSE.          *)
(*                                                                        *)
(**************************************************************************)
(* Adapted significantly by ReScript Authors                          *)
module A = Belt_Array

type 'a node = { 
  content: 'a; 
  mutable next: 'a cell 
}
and 'a cell = 'a node option 
and 'a t = {
  mutable length: int;
  mutable first: 'a cell;
  mutable last: 'a cell
} 




let make () =   
    {
      length = 0;
      first = None;
      last = None }


let clear q =
  q.length <-  0;
  q.first <-  None;
  q.last <-  None

let add q x =
  let cell = Some {
      content = x;
      next = None } in
  match  q.last with
  | None -> (* TODO: better names for intermediate var *)
    q.length <- 1;
    q.first <- cell;
    q.last <- cell
  | Some last ->
    q.length <- q.length + 1;
    last.next <-  cell;
    q.last <-  cell


  
let peek q =
  match  q.first with (* same here could be v *)
  | None -> None
  | Some v -> Some v.content

let peekUndefined q =
  match  q.first with
  | None -> Js.undefined
  | Some v -> Js.Undefined.return v.content


let peekExn q =
  match  q.first with
  | None -> raise Not_found
  | Some v -> v.content

let pop q =
  match  q.first with
  | None -> None
  | Some x  ->
    let next = x.next in 
    if next = None then 
      begin (* only one element*)
        clear q;
        Some x.content
      end
    else begin 
      q.length <- q.length  - 1;
      q.first <- next;
      Some x.content 
    end

let popExn q = (* TO fix *)
  match  q.first with 
  | None -> raise Not_found
  | Some x  ->
    let next = x.next in 
    if next = None then 
      begin (* only one element*)
        clear q;
        x.content
      end
    else begin 
      q.length <- q.length - 1;
      q.first <- next;
      x.content 
    end    

let popUndefined q =
  match  q.first with
  | None -> Js.undefined
  | Some x  ->
    let next = x.next in 
    if next = None then 
      begin (* only one element*)
        clear q;
        Js.Undefined.return x.content
      end
    else begin 
      q.length <- q.length - 1;
      q.first <- next;
      Js.Undefined.return x.content 
    end

let rec copyAux qRes prev cell =
  match  cell with
  | None -> qRes.last <-  prev; qRes
  | Some x  ->
    let content = x.content in 
    let res =  Some {content ; next = None}  in
    begin match  prev with
      | None ->  qRes.first <- res
      | Some p -> p.next <-  res
    end;
    copyAux qRes res (x.next)

let copy q =
  copyAux {length = q.length; first = None;  last = None}  None q.first


      
let rec copyMapAux qRes prev cell f =
  match  cell with
  | None ->  qRes.last <- prev; qRes
  | Some x  ->
    let content = f x.content [@bs] in 
    let res = Some {content; next = None}  in
    begin match  prev with (*TODO: optimize to remove such check*)
      | None -> qRes.first <- res
      | Some p -> p.next <-  res
    end;
    copyMapAux qRes res x.next f

let mapU q f =
  copyMapAux {length = q.length; first = None; last = None}  None q.first f
    
let map q f = mapU q (fun [@bs] a -> f a)
    
let isEmpty q =
  q.length = 0

let size q =
  q.length

let rec iterAux cell f =
  match  cell with
  | None -> ()
  | Some x  ->
    f x.content [@bs];
    iterAux (x.next) f

let forEachU q f =
  iterAux q.first f

let forEach q f = forEachU q (fun[@bs] a -> f a)
    
let rec foldAux f accu cell =
  match  cell with
  | None -> accu
  | Some x  ->
    let accu = f accu x.content [@bs] in
    foldAux f accu (x.next)

let reduceU q  accu f =
  foldAux f accu q.first

let reduce q accu f = reduceU q accu (fun[@bs] a b -> f a b)
    
let transfer q1 q2 =
  if q1.length > 0 then
    match  q2.last with
    | None ->
      q2.length <- q1.length;
      q2.first <- q1.first;
      q2.last <- q1.last;
      clear q1
    | Some l ->
      q2.length <- (q2.length + q1.length);
      l.next <- q1.first;
      q2.last <- q1.last;
      clear q1

let rec fillAux i arr cell =       
  match  cell with 
  | None -> ()
  | Some x ->
    A.setUnsafe arr i x.content;
    fillAux (i + 1) arr (x.next) 

let toArray x =         
  let v = A.makeUninitializedUnsafe x.length in 
  fillAux 0 v x.first;
  v

(*TODO: optimize *)
let fromArray arr =
  let q = make () in
  for i = 0 to A.length arr - 1 do
    add q (A.getUnsafe arr i)
  done ;
  q

File: lib/es6/belt_MutableQueue.mjs


import * as Curry from "./curry.mjs";
import * as Caml_option from "./caml_option.mjs";

function make(param) {
  return {
          length: 0,
          first: undefined,
          last: undefined
        };
}

function clear(q) {
  q.length = 0;
  q.first = undefined;
  q.last = undefined;
  
}

function add(q, x) {
  var cell = {
    content: x,
    next: undefined
  };
  var last = q.last;
  if (last !== undefined) {
    q.length = q.length + 1 | 0;
    last.next = cell;
    q.last = cell;
  } else {
    q.length = 1;
    q.first = cell;
    q.last = cell;
  }
  
}

function peek(q) {
  var v = q.first;
  if (v !== undefined) {
    return Caml_option.some(v.content);
  }
  
}

function peekUndefined(q) {
  var v = q.first;
  if (v !== undefined) {
    return v.content;
  }
  
}

function peekExn(q) {
  var v = q.first;
  if (v !== undefined) {
    return v.content;
  }
  throw {
        RE_EXN_ID: "Not_found",
        Error: new Error()
      };
}

function pop(q) {
  var x = q.first;
  if (x === undefined) {
    return ;
  }
  var next = x.next;
  if (next === undefined) {
    clear(q);
    return Caml_option.some(x.content);
  } else {
    q.length = q.length - 1 | 0;
    q.first = next;
    return Caml_option.some(x.content);
  }
}

function popExn(q) {
  var x = q.first;
  if (x !== undefined) {
    var next = x.next;
    if (next === undefined) {
      clear(q);
      return x.content;
    } else {
      q.length = q.length - 1 | 0;
      q.first = next;
      return x.content;
    }
  }
  throw {
        RE_EXN_ID: "Not_found",
        Error: new Error()
      };
}

function popUndefined(q) {
  var x = q.first;
  if (x === undefined) {
    return ;
  }
  var next = x.next;
  if (next === undefined) {
    clear(q);
    return x.content;
  } else {
    q.length = q.length - 1 | 0;
    q.first = next;
    return x.content;
  }
}

function copy(q) {
  var qRes = {
    length: q.length,
    first: undefined,
    last: undefined
  };
  var _prev;
  var _cell = q.first;
  while(true) {
    var cell = _cell;
    var prev = _prev;
    if (cell !== undefined) {
      var content = cell.content;
      var res = {
        content: content,
        next: undefined
      };
      if (prev !== undefined) {
        prev.next = res;
      } else {
        qRes.first = res;
      }
      _cell = cell.next;
      _prev = res;
      continue ;
    }
    qRes.last = prev;
    return qRes;
  };
}

function mapU(q, f) {
  var qRes = {
    length: q.length,
    first: undefined,
    last: undefined
  };
  var _prev;
  var _cell = q.first;
  while(true) {
    var cell = _cell;
    var prev = _prev;
    if (cell !== undefined) {
      var content = f(cell.content);
      var res = {
        content: content,
        next: undefined
      };
      if (prev !== undefined) {
        prev.next = res;
      } else {
        qRes.first = res;
      }
      _cell = cell.next;
      _prev = res;
      continue ;
    }
    qRes.last = prev;
    return qRes;
  };
}

function map(q, f) {
  return mapU(q, Curry.__1(f));
}

function isEmpty(q) {
  return q.length === 0;
}

function size(q) {
  return q.length;
}

function forEachU(q, f) {
  var _cell = q.first;
  while(true) {
    var cell = _cell;
    if (cell === undefined) {
      return ;
    }
    f(cell.content);
    _cell = cell.next;
    continue ;
  };
}

function forEach(q, f) {
  return forEachU(q, Curry.__1(f));
}

function reduceU(q, accu, f) {
  var _accu = accu;
  var _cell = q.first;
  while(true) {
    var cell = _cell;
    var accu$1 = _accu;
    if (cell === undefined) {
      return accu$1;
    }
    var accu$2 = f(accu$1, cell.content);
    _cell = cell.next;
    _accu = accu$2;
    continue ;
  };
}

function reduce(q, accu, f) {
  return reduceU(q, accu, Curry.__2(f));
}

function transfer(q1, q2) {
  if (q1.length <= 0) {
    return ;
  }
  var l = q2.last;
  if (l !== undefined) {
    q2.length = q2.length + q1.length | 0;
    l.next = q1.first;
    q2.last = q1.last;
    return clear(q1);
  } else {
    q2.length = q1.length;
    q2.first = q1.first;
    q2.last = q1.last;
    return clear(q1);
  }
}

function fillAux(_i, arr, _cell) {
  while(true) {
    var cell = _cell;
    var i = _i;
    if (cell === undefined) {
      return ;
    }
    arr[i] = cell.content;
    _cell = cell.next;
    _i = i + 1 | 0;
    continue ;
  };
}

function toArray(x) {
  var v = new Array(x.length);
  fillAux(0, v, x.first);
  return v;
}

function fromArray(arr) {
  var q = {
    length: 0,
    first: undefined,
    last: undefined
  };
  for(var i = 0 ,i_finish = arr.length; i < i_finish; ++i){
    add(q, arr[i]);
  }
  return q;
}

export {
  make ,
  clear ,
  isEmpty ,
  fromArray ,
  add ,
  peek ,
  peekUndefined ,
  peekExn ,
  pop ,
  popUndefined ,
  popExn ,
  copy ,
  size ,
  mapU ,
  map ,
  forEachU ,
  forEach ,
  reduceU ,
  reduce ,
  transfer ,
  toArray ,
  
}
/* No side effect */
